src/metabase/pulse/render/body.clj

```diff
@@ -109,13 +124,26 @@
               :when remapped_from]
           [remapped_from col-idx])))
 
+(defn- ref-key-with-id?
+  "True if the `settings-key` is a ref and matches the id. The key will look something like:
+  [\"ref\" [\"field-ref\" id {:maybe more}]]"
+  [settings-key id]
+  (and (= "ref" (first settings-key))
+       (= id (get-in settings-key [1 1]))))
+
 (defn- column-name
   "Returns first column name from a hierarchy of possible column names"
   [card col]
   (let [column-settings (some->> (get-in card [:visualization_settings :column_settings])
                                  (m/map-keys (comp vec json/parse-string name)))]
-    (name (or (when-let [fr (:field_ref col)]
-                (get-in column-settings [["ref" (mapv #(if (keyword? %) (name %) %) fr)] :column_title]))
+    (name (or (when-let [[_ id _] (:field_ref col)]
+                ;; field-ref keys can come in with additional stuff like :meta-data or unit maps,
+                ;; so we just use the field_ref ID number to try find the settings for this particular column
+                (->> column-settings
+                     (filter #(ref-key-with-id? (first %) id))
+                     first
+                     second
+                     :column_title))
```

Comment:
Clearly I'm new to this code, but this is another case where using Schema would help future readers understand the shape of the data (never mind the extra assurance of correctness)